tools/eve-installer-gen.sh

```diff
@@ -0,0 +1,138 @@
+     Production EVE versions are in the format: <major>.<minor>.<bugfix>
+     Mandatory argument.
+  -a <target device architecture>
+     Should be one of: amd64 | arm64 | riscv64
+     Mandatory argument.
+  -s <soft serial string>
+  -c <path to onboarding certificate>
+     The certificate should be in the PEM format.
+  -k <path to onboarding key>
+     The key should be in the PEM format.
+  -b <path to bootstrap config>
+     The file should contain BootstrapConfig (see EVE API) serialized with protobuf binary encoding.
+  -o <path where to store output EVE installer>
+     Mandatory argument.
+  -f <output image format>
+     Should be one of: raw | iso
+     Default is raw.
+Examples:
+  $0 -u zedcloud.mycluster.zededa.net -v 8.5.1 -a amd64 -s 123456789 -b ./config.pb -o installer.raw
+"
+
+bail() {
+  [ -n "$1" ] && echo "$1" >&2
+  echo "$USAGE" >&2
+  exit 1
+}
+
+cleanup() {
+  [ -n "$CONTAINER" ] && docker rm "$CONTAINER" >/dev/null 2>&1
+  [ -n "$VOLUME" ] && docker volume rm "$VOLUME" >/dev/null 2>&1
+  [ -n "$TMPDIR" ] && rm -rf "$TMPDIR"
+}
+
+while getopts u:v:a:s:c:k:b:o:f: o
+do case "$o" in
+  u)   CONTROLLER="$OPTARG";;
+  v)   VERSION="$OPTARG";;
+  a)   ARCH="$OPTARG";;
+  s)   SERIAL="$OPTARG";;
+  c)   CERT="$OPTARG";;
+  k)   KEY="$OPTARG";;
+  b)   BOOTSTRAP="$OPTARG";;
+  o)   OUTPUT="$OPTARG";;
+  f)   FORMAT="$OPTARG";;
+  [?]) bail;;
+  esac
+done
+
+if [ -z "$VERSION" ]; then
+  bail "Missing EVE version argument (-v)"
+fi
+
+if [ -z "$ARCH" ]; then
+  bail "Missing arch argument (-a)"
+fi
+
+if [ -z "$OUTPUT" ]; then
+  bail "Missing output argument (-o)"
+fi
+
+case "$ARCH" in
+  "amd64");;
+  "arm64");;
+  "riscv64");;
+  *) bail "Unrecognized architecture ($ARCH)";;
+esac
+
+case "$FORMAT" in
+  "") FORMAT="raw";; # default value
+  "raw");;
+  "iso");;
+  *) bail "Unrecognized output format ($FORMAT)";;
+esac
+
+if [ -n "$CERT" ] && [ ! -f "$CERT" ]; then
+  bail "Onboarding certificate file ($CERT) does not exist"
+fi
+
+if [ -n "$CERT" ] && [ -z "$KEY" ]; then
+  bail "Onboarding certificate defined but not the key"
+fi
+
+if [ -n "$KEY" ] && [ ! -f "$KEY" ]; then
+  bail "Onboarding key file ($KEY) does not exist"
+fi
+
+if [ -n "$BOOTSTRAP" ] && [ ! -f "$BOOTSTRAP" ]; then
+  bail "Bootstrap config file ($BOOTSTRAP) does not exist"
+fi
+
+TMPDIR="$(mktemp -d -t config-override-XXXXXXXXXX)"
+if [ -n "$CONTROLLER" ]; then
+  echo "$CONTROLLER" > "$TMPDIR/server"
+  FILES="$FILES server"
+fi
+if [ -n "$SERIAL" ]; then
+  echo "$SERIAL" > "$TMPDIR/soft_serial"
+  FILES="$FILES soft_serial"
+fi
+if [ -n "$CERT" ]; then
+  cp "$CERT" "$TMPDIR/onboard.cert.pem"
+  cp "$KEY" "$TMPDIR/onboard.key.pem"
+  FILES="$FILES onboard.cert.pem onboard.key.pem"
+fi
+if [ -n "$BOOTSTRAP" ]; then
+  cp "$BOOTSTRAP" "$TMPDIR/bootstrap-config.pb"
+  FILES="$FILES bootstrap-config.pb"
+fi
+
+# Prepare lfedge/eve docker container with config files to override/add before
+# starting the image builder.
+TAG="$VERSION-kvm-$ARCH"
```

Comment:
Instead of hardcoding the kvm part would it make sense to have the version string be x.y.z.<hyp>.<arch>?

Reply:
Or I could add `-h <hypervisor>` argument. However, I'm not sure if we can rely on EVE image tag always following the `<version>-<hypervisor>-<arch>` pattern (say images built outside of lfedge, e.g. by OEMs). So in that regard, requiring a full eve image tag as one argument might be safer.

Reply:
I decided to go with `-t <EVE image tag>` (i.e. not making assumptions about image tag format).
Also I decided to allocate "c" arg for the controller address and use "e" for onboarding certificate opt.